audit/hashstructure.go

```diff
@@ -187,11 +193,19 @@ func HashWrapInfo(salter *salt.Salt, in *wrapping.ResponseWrapInfo, HMACAccessor
 
 // HashStructure takes an interface and hashes all the values within
 // the structure. Only _values_ are hashed: keys of objects are not.
+// Note that the function takes both the original data structure and
+// an unmarshaled json copy.  The hashed values are written to the copy
+// but because the json transform has changed the type info the original
+// must also be walked to get that type info.
+//
+// The original is walked with the reflectwalk.Walk() method below.
+// That method leaves a trail in w.loc[]/w.csKey[] which is then used
+// by getValueFromCopy() to walk the copy.
 //
 // For the HashCallback, see the built-in HashCallbacks below.
-func HashStructure(s interface{}, cb HashCallback, ignoredKeys []string) error {
-	walker := &hashWalker{Callback: cb, IgnoredKeys: ignoredKeys}
-	return reflectwalk.Walk(s, walker)
+func HashStructure(original interface{}, copy interface{}, cb HashCallback, ignoredKeys []string, elideListResponseData bool) error {
+	walker := &hashWalker{UnmarshalledCopy: reflect.ValueOf(copy), Callback: cb, IgnoredKeys: ignoredKeys, ElideListResponseData: elideListResponseData}
```

Comment:
My only comment here is that I wonder if we can walk `copy` directly? Since it is the output of JSON unmarshaling into `map[string]interface{}`, we know the expected structure and can directly index it with type assertions, always to `map[string]interface{}` by virtue of how the JSON deserialization works.

Reply:
I guess we'd need to reflect at the very last step of `getValueFromCopy` anyways, so probably not worth it.

Reply:
walking directly might be a good idea.  (it didn't occur to me to be honest.)

Let me play with it a bit and see.

Reply:
@cipherboy I changed it to do direct writes.  I'm not sure it is any better, but you can take a look.